src/imp_std.rs

```diff
@@ -76,131 +70,140 @@ impl<T> OnceCell<T> {
-    /// successful write.
     #[cold]
     pub(crate) fn initialize<F, E>(&self, f: F) -> Result<(), E>
     where
         F: FnOnce() -> Result<T, E>,
     {
-        let mut f = Some(f);
-        let mut res: Result<(), E> = Ok(());
-        let slot = &self.value;
-        initialize_inner(&self.state, &mut || {
-            let f = f.take().unwrap();
-            match f() {
-                Ok(value) => {
-                    unsafe { *slot.get() = Some(value) };
-                    true
+        // Note on atomic orderings of `self.status`:
+        // - The only data that has to be synchronised across threads is `self.value`.
+        // - The only store that needs `Release` is after the data is written and `status` is set to
+        //   COMPLETE (in the destructor of `wake_waiters_on_drop`).
+        //   All other stores can be `Relaxed`.
+        // - At the end of this `initialize` function we have to guarantee the data is acquired.
+        //   Every load that can lead to the end of this function must have `Acquire` ordering.
+        let mut state_and_queue = self.state_and_queue.load(Ordering::Acquire);
+        loop {
+            match state_and_queue {
+                COMPLETE => break,
+                EMPTY => {
+                    // Try to register this thread as the one doing initialization (RUNNING).
+                    let old = self.state_and_queue.compare_and_swap(EMPTY,
+                                                                    RUNNING,
+                                                                    Ordering::Acquire);
+                    if old != EMPTY {
+                        state_and_queue = old;
+                        continue;
+                    }
+                    // The destructor of `wake_waiters_on_drop` will do cleanup here.
+                    // It will wake up other threads that may have waited on us.
+                    // If the closure panics or returns Err it will reset `state_and_queue` to
+                    // EMPTY, otherwise to COMPLETED.
+                    let mut wake_waiters_on_drop = WaiterQueue::new(&self.state_and_queue);
+                    // Run the initialization closure.
+                    let value = f()?;
+                    let slot: &mut Option<T> = unsafe { &mut *self.value.get() };
+                    debug_assert!(slot.is_none());
+                    *slot = Some(value);
+                    wake_waiters_on_drop.completed = true;
+                    break;
                 }
-                Err(e) => {
-                    res = Err(e);
-                    false
+                _ => {
+                    wait(&self.state_and_queue, state_and_queue);
+                    state_and_queue = self.state_and_queue.load(Ordering::Acquire);
                 }
             }
-        });
-        res
+        }
+        Ok(())
     }
 }
 
-// Note: this is intentionally monomorphic
```

Comment:
I was not sure what to make of this

Reply:
Yeah, I should have written a better comment here. 

Basically, this function is not generic, which means that we only compile it once, instead of for each subsitution of generic parameters of OnceCell in each crate. That is, using `&mut dyn` instead of `&mut FnOnce` helps with compile times and reduces unnecessary code bloat

Reply:
The `parking_lot` variant does not have this property, but that seems useful to add back here.